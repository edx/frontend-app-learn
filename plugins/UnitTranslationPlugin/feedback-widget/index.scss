.action-row-divider {
    font-size: 31px;
    font-weight: 100;
}