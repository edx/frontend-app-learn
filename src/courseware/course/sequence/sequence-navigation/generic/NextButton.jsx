import PropTypes from 'prop-types';
import { Link, useLocation } from 'react-router-dom';
import { Button } from '@openedx/paragon';
import { ChevronLeft, ChevronRight } from '@openedx/paragon/icons';
import { isRtl, getLocale } from '@edx/frontend-platform/i18n';

import UnitNavigationEffortEstimate from '../UnitNavigationEffortEstimate';

const NextButton = ({
  onClickHandler,
  buttonText,
  nextLink,
  variant,
  buttonStyle,
  disabled,
  hasEffortEstimate,
}) => {
  const nextArrow = isRtl(getLocale()) ? ChevronLeft : ChevronRight;
  const { pathname } = useLocation();
  const navLink = pathname.startsWith('/preview') ? `/preview${nextLink}` : nextLink;
  const buttonContent = hasEffortEstimate ? (
    <UnitNavigationEffortEstimate>
      {buttonText}
    </UnitNavigationEffortEstimate>
  ) : buttonText;

  return (
    <Button
      variant={variant}
      className={buttonStyle}
      disabled={disabled}
      onClick={onClickHandler}
      as={disabled ? undefined : Link}
      to={disabled ? undefined : navLink}
      iconAfter={nextArrow}
    >
      {buttonContent}
    </Button>
  );
};

NextButton.defaultProps = {
  hasEffortEstimate: false,
};

NextButton.propTypes = {
  onClickHandler: PropTypes.func.isRequired,
  buttonText: PropTypes.string.isRequired,
  nextLink: PropTypes.string.isRequired,
  variant: PropTypes.string.isRequired,
  buttonStyle: PropTypes.string.isRequired,
  disabled: PropTypes.bool.isRequired,
  hasEffortEstimate: PropTypes.bool,
};

export default NextButton;
