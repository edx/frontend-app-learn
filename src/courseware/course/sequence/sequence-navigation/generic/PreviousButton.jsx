import PropTypes from 'prop-types';
import { Link, useLocation } from 'react-router-dom';
import { Button } from '@openedx/paragon';
import { ChevronLeft, ChevronRight } from '@openedx/paragon/icons';
import { isRtl, getLocale } from '@edx/frontend-platform/i18n';

const PreviousButton = ({
  onClick,
  buttonLabel,
  previousLink,
  variant,
  buttonStyle,
  isFirstUnit,
}) => {
  const disabled = isFirstUnit;
  const prevArrow = isRtl(getLocale()) ? ChevronRight : ChevronLeft;
  const { pathname } = useLocation();
  const navLink = pathname.startsWith('/preview') ? `/preview${previousLink}` : previousLink;

  return (
    <Button
      variant={variant}
      className={buttonStyle}
      disabled={disabled}
      onClick={onClick}
      as={disabled ? undefined : Link}
      to={disabled ? undefined : navLink}
      iconBefore={prevArrow}
    >
      {buttonLabel}
    </Button>
  );
};

PreviousButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  buttonLabel: PropTypes.string.isRequired,
  previousLink: PropTypes.string.isRequired,
  variant: PropTypes.string.isRequired,
  buttonStyle: PropTypes.string.isRequired,
  isFirstUnit: PropTypes.bool.isRequired,
};

export default PreviousButton;
