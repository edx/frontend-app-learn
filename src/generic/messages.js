import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  close: {
    id: 'general.altText.close',
    defaultMessage: 'Close',
    description: 'Text used as an aria-label to describe closing or dismissing a component',
  },
});

export default messages;
