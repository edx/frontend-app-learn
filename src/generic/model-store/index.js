export {
  reducer,
  addModel,
  addModels,
  addModelsMap,
  updateModel,
  updateModels,
  updateModelsMap,
  removeModel,
  removeModels,
  setSearchEngineAuthToken,
} from './slice';

export {
  useModel,
  useModels,
} from './hooks';
