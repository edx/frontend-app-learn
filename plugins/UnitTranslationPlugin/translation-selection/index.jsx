import React, { useContext, useEffect } from 'react';
import PropTypes from 'prop-types';

import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import { AppContext } from '@edx/frontend-platform/react';
import { IconButton, Icon, ProductTour } from '@openedx/paragon';
import { Language } from '@openedx/paragon/icons';
import { useDispatch } from 'react-redux';
import { stringifyUrl } from 'query-string';

import { registerOverrideMethod } from '@src/generic/plugin-store';

import TranslationModal from './TranslationModal';
import useTranslationTour from './useTranslationTour';
import useSelectLanguage from './useSelectLanguage';
import FeedbackWidget from '../feedback-widget';

const TranslationSelection = ({
  id, courseId, language, availableLanguages, unitId,
}) => {
  const {
    authenticatedUser: { userId },
  } = useContext(AppContext);
  const dispatch = useDispatch();
  const {
    translationTour, isOpen, open, close,
  } = useTranslationTour();

  const { selectedLanguage, setSelectedLanguage } = useSelectLanguage({
    courseId,
    language,
  });

  useEffect(() => {
    if ((courseId && language && selectedLanguage && unitId && userId) && (language !== selectedLanguage)) {
      const eventName = 'edx.whole_course_translations.translation_requested';
      const eventProperties = {
        courseId,
        sourceLanguage: language,
        targetLanguage: selectedLanguage,
        unitId,
        userId,
      };
      sendTrackEvent(eventName, eventProperties);
    }
  }, [courseId, language, selectedLanguage, unitId, userId]);

  useEffect(() => {
    dispatch(
      registerOverrideMethod({
        pluginName: id,
        methodName: 'getIFrameUrl',
        method: (iframeUrl) => {
          const finalUrl = stringifyUrl({
            url: iframeUrl,
            query: {
              ...(language
                && selectedLanguage
                && language !== selectedLanguage && {
                src_lang: language,
                dest_lang: selectedLanguage,
              }),
            },
          });
          return finalUrl;
        },
      }),
    );
  }, [language, selectedLanguage]);

  return (
    <>
      <ProductTour tours={[translationTour]} />
      <IconButton
        src={Language}
        iconAs={Icon}
        alt="change-language"
        onClick={open}
        variant="primary"
        className="mr-2 mb-2 float-right"
        id="translation-selection-button"
      />
      <TranslationModal
        isOpen={isOpen}
        close={close}
        courseId={courseId}
        selectedLanguage={selectedLanguage}
        setSelectedLanguage={setSelectedLanguage}
        availableLanguages={availableLanguages}
        id={id}
      />
      <FeedbackWidget
        courseId={courseId}
        translationLanguage={selectedLanguage}
        unitId={unitId}
        userId={userId}
      />
    </>
  );
};

TranslationSelection.propTypes = {
  id: PropTypes.string.isRequired,
  courseId: PropTypes.string.isRequired,
  unitId: PropTypes.string.isRequired,
  language: PropTypes.string.isRequired,
  availableLanguages: PropTypes.arrayOf(PropTypes.shape({
    code: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
  })).isRequired,
};

TranslationSelection.defaultProps = {};

export default TranslationSelection;
