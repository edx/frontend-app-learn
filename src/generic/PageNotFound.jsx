import { getConfig } from '@edx/frontend-platform';
import { Hyperlink } from '@openedx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';
import FooterSlot from '@openedx/frontend-slot-footer';

import HeaderSlot from '../plugin-slots/HeaderSlot';
import messages from './messages';

const PageNotFound = () => {
  const { formatMessage } = useIntl();

  return (
    <>
      <HeaderSlot />
      <main
        id="main-content"
        className="main-content d-flex justify-content-center align-items-center flex-column"
        style={{
          height: '50vh',
        }}
      >
        <h1 className="h3">
          {formatMessage(messages.pageNotFoundHeader)}
        </h1>
        <p>
          {formatMessage(
            messages.pageNotFoundBody,
            {
              homepageLink: (
                <Hyperlink destination={getConfig().LMS_BASE_URL}>
                  {formatMessage(messages.homepageLink)}
                </Hyperlink>
              ),
            },
          )}
        </p>
      </main>
      <FooterSlot />
    </>
  );
};

export default PageNotFound;
