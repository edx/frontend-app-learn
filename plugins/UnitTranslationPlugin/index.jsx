import React from 'react';
import PropTypes from 'prop-types';

import { useModel } from '@src/generic/model-store';

import TranslationSelection from './translation-selection';

const UnitTranslationPlugin = ({ id, courseId }) => {
  const { language, wholeCourseTranslationEnabled } = useModel(
    'courseHomeMeta',
    courseId,
  );

  if (!wholeCourseTranslationEnabled || !language) {
    return null;
  }

  return <TranslationSelection id={id} courseId={courseId} language={language} />;
};

UnitTranslationPlugin.propTypes = {
  id: PropTypes.string.isRequired,
  courseId: PropTypes.string.isRequired,
};

export default UnitTranslationPlugin;
