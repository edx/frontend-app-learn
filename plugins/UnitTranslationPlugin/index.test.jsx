import { shallow } from '@edx/react-unit-test-utils';
import { useState } from 'react';
import { useModel } from '@src/generic/model-store';

import UnitTranslationPlugin from './index';

jest.mock('@src/generic/model-store');
jest.mock('./data/api', () => ({
  fetchTranslationConfig: jest.fn(),
}));
jest.mock('./translation-selection', () => 'TranslationSelection');
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

describe('<UnitTranslationPlugin />', () => {
  const props = {
    id: 'id',
    courseId: 'courseId',
    unitId: 'unitId',
  };
  const mockInitialState = ({ enabled = true, availableLanguages = ['en'] }) => {
    useState.mockReturnValue([{ enabled, availableLanguages }, jest.fn()]);
  };
  it('render empty when translation is not enabled', () => {
    useModel.mockReturnValue({ language: 'en' });
    mockInitialState({ enabled: false });

    const wrapper = shallow(<UnitTranslationPlugin {...props} />);

    expect(wrapper.isEmptyRender()).toBe(true);
  });
  it('render empty when available languages is empty', () => {
    useModel.mockReturnValue({ language: 'fr' });
    mockInitialState({
      availableLanguages: [],
    });

    const wrapper = shallow(<UnitTranslationPlugin {...props} />);

    expect(wrapper.isEmptyRender()).toBe(true);
  });

  it('render empty when course language has not been set', () => {
    useModel.mockReturnValue({ language: undefined });
    mockInitialState({});

    const wrapper = shallow(<UnitTranslationPlugin {...props} />);

    expect(wrapper.isEmptyRender()).toBe(true);
  });

  it('render TranslationSelection when translation is enabled and language is available', () => {
    useModel.mockReturnValue({ language: 'en' });
    mockInitialState({});

    const wrapper = shallow(<UnitTranslationPlugin {...props} />);

    expect(wrapper.snapshot).toMatchSnapshot();
  });
});
