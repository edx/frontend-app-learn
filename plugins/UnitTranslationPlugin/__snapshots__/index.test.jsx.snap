// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UnitTranslationPlugin /> render TranslationSelection when translation is enabled and language is available 1`] = `
<TranslationSelection
  availableLanguages={
    Array [
      "en",
    ]
  }
  courseId="courseId"
  id="id"
  language="en"
  unitId="unitId"
/>
`;
